src/Microsoft.Kusto.ServiceLayer/DataSource/KustoDataSource.cs

```diff
@@ -626,93 +612,311 @@ internal IEnumerable<ColumnInfo> GetColumnMetadata(string databaseName)
+        }
+
+        private IEnumerable<FunctionInfo> GetFunctionInfos(string databaseName)
+        {
+            CancellationTokenSource source = new CancellationTokenSource();
+            CancellationToken token = source.Token;
 
-                // Fill in all column metadata
-                tableGroups.ToList().ForEach(grouping => {
-                    columnMetadata[GenerateColumnMetadataKey(databaseName, grouping.Key)] = grouping
-                        .Select(row => new ColumnMetadata
-                            {
-                                ClusterName = this.ClusterName,
-                                DatabaseName = databaseName,
-                                TableName = grouping.Key,
-                                MetadataType = DataSourceMetadataType.Column,
-                                MetadataTypeName = DataSourceMetadataType.Column.ToString(),
-                                Name = row.Name, // column name
-                                PrettyName = row.Name,
-                                Urn = $"{this.ClusterName}.{databaseName}.{grouping.Key}.{row.Name}",
-                                DataType = row.DataType
-                            })
-                        .OrderBy(row => row.Name, StringComparer.OrdinalIgnoreCase); // case-insensitive
-                });
-            }
-        }
-
-        private string GenerateColumnMetadataKey(string databaseName, string tableName)
-        {
-            return $"{databaseName}.{tableName}";
+            string query = ".show functions";
+
+            using (var reader = ExecuteQuery(query, token, databaseName))
+            {
+                return reader.ToEnumerable()
+                    .Select(row => new FunctionInfo
+                    {
+                        Name = row["Name"]?.ToString(),
+                        Body = row["Body"]?.ToString(),
+                        DocString = row["DocString"]?.ToString(),
+                        Folder = row["Folder"]?.ToString(),
+                        Parameters = row["Parameters"]?.ToString()
+                    })
+                    .Materialize();
+            }
+        }
+
+        internal FunctionInfo GetFunctionInfo(string functionName)
+        {
+            CancellationTokenSource source = new CancellationTokenSource();
+            CancellationToken token = source.Token;
+
+            string query = $".show function {functionName}";
+
+            using (var reader = ExecuteQuery(query, token, DatabaseName))
+            {
+                return reader.ToEnumerable()
+                    .Select(row => new FunctionInfo
+                    {
+                        Name = row["Name"]?.ToString(),
+                        Body = row["Body"]?.ToString(),
+                        DocString = row["DocString"]?.ToString(),
+                        Folder = row["Folder"]?.ToString(),
+                        Parameters = row["Parameters"]?.ToString()
+                    })
```

Comment:
I wonder if FunctionInfo and FunctionObjectMetadata,  and ColumnInfo and ColumnObjectMetadata can be combined into one. These 2 pairs seem to be very similar. It is ok if you leave this as an exercise for the  future. Just sharing the thought here.

Reply:
It's a good thought. Atm I was keeping the models separate. The Info is an internal model from the query results to be passed around the datasource. The Metadata is a model for ADS.